### PR TITLE
Keep pw based auth tokens valid when pw-less login happens

### PR DESCRIPTION
**Pull request by 7118path**
Hi,

after publishing the eID-Login app for Nextcloud users reported about the problem, that after doing a passwordless login with an eID, all already existing password based authn tokens (browser sessions and app passwords) become invalidated. See eid-login/eid-login-nextcloud#2

When investigating the issue, I found the same happens when doing a WebAuthn (password ) login. The empty password from the login is used to update existing authn tokens, which invalidates them. They get removed on the next run of https://github.com/nextcloud/server/blob/c2f62ee0b8c779808992d053494a845daa3d33b7/lib/private/User/Session.php#L729

To keep the existing authn tokens valid, this pull requests prevents the updating of the tokens, if the given password is an empty string.

Please note: This are the two places I found where token passwords are updated on an login attempt. I am not sure why this must happen twice? There may be more? I don`t know.

Also there is a little inconsistency in the function signatures for setting token passwords. When creating a session token a 'null' value as password can be set, which is also set it as 'null' in the database: 
https://github.com/nextcloud/server/blob/c2f62ee0b8c779808992d053494a845daa3d33b7/lib/private/User/Session.php#L679
This is resulting in different handling of the token later on:
https://github.com/nextcloud/server/blob/c4159e37ed594030b5be630139426c05a07c158e/lib/private/Authentication/Token/DefaultTokenProvider.php#L224-L230
https://github.com/nextcloud/server/blob/c2f62ee0b8c779808992d053494a845daa3d33b7/lib/private/User/Session.php#L740-L748

In these functions 'null' is not allowed but an empty string must be used:
https://github.com/nextcloud/server/blob/c4159e37ed594030b5be630139426c05a07c158e/lib/private/Authentication/Token/DefaultTokenProvider.php#L86
https://github.com/nextcloud/server/blob/c4159e37ed594030b5be630139426c05a07c158e/lib/private/Authentication/Token/DefaultTokenProvider.php#L242

Maybe this APIs can be changed to handle the 'null' value more consistently?

TIA for looking into this and feedback about the issue!

Greetings!

### CODE REVIEW

File: lib/private/Authentication/Token/PublicKeyTokenProvider.php [hunk 1]
```diff
@@ -414,6 +414,11 @@ public function markPasswordInvalid(IToken $token, string $tokenId) {
 	public function updatePasswords(string $uid, string $password) {
 		$this->cache->clear();
 
+		// prevent setting an empty pw as result of pw-less-login
+		if ($password==='') {
```
Comment (nickvergessen) [inline lib/private/Authentication/Token/PublicKeyTokenProvider.php:418]:
```suggestion
		if ($password === '') {
```

File: lib/private/Authentication/Listeners/UserLoggedInListener.php [hunk 2]
```diff
@@ -48,6 +48,11 @@ public function handle(Event $event): void {
 			return;
 		}
 
+		// prevent setting an empty pw as result of pw-less-login
+		if ($event->getPassword()==='') {
```
Comment (nickvergessen) [inline lib/private/Authentication/Listeners/UserLoggedInListener.php:52]:
```suggestion
		if ($event->getPassword() === '') {
```

**Review by nickvergessen — APPROVED** [review]

**Comment by nickvergessen** [comment]
> Conflicting files
lib/private/Authentication/Token/PublicKeyTokenProvider.php

You unluckily need to rebase on latest master

**Review by juliusknorr — APPROVED** [review]

**Comment by juliushaertl** [comment]
@ecsecta Could you do a rebase of the branch instead of a merge with master and maybe also squash your commits into one?

**Comment by ecsecta** [comment]
@juliushaertl The last commit should contain all changes based on the latest master. Merging this commit should be sufficient. I have no idea how to squash all commit, when not merging. If I am wrong, maybe you could point me to some docs how to do what is needed now?

**Review by LukasReschke — APPROVED** [review]

**Comment by ecsecta** [comment]
Thank you!

**Comment by ChristophWurst** [comment]
more lucky on Friday 13th? let's fine out

**Pull request merged**